src/utf8.c

```diff
@@ -0,0 +1,435 @@
+			               "\\\\x%02x-\\\\x%02x", a, a + b - 1);
+			if (n < 1)
+				return;
+			w += (size_t)n;
+		}
+		if (d) {
+			n = d == 1
+			    ? snprintf(&buf[w], sizeof buf - w,
+			               "\\\\x%02x", a + b + c)
+			    : snprintf(&buf[w], sizeof buf - w,
+			               "\\\\x%02x-\\\\x%02x", a + b + c, a + b + c + d - 1);
+			if (n < 1)
+				return;
+			w += (size_t)n;
+		}
+		buf[w++] = ']';
+		buf[w++] = '\"';
+		buf[w++] = ']';
+		buf[w++] = ';';
+		buf[w++] = '\n';
+	}
+
+	for (size_t src = 0;
+	     src < sizeof utf8_dst /
+	           sizeof utf8_dst[0]; ++src) {
+		if (!(utf8_range[src][1] +
+		      utf8_range[src][3]))
+			continue;
+		for (unsigned bits = utf8_dst[src],
+		     dst = 0; bits; bits >>= 1U, ++dst) {
+			if (!(bits & 1U))
+				continue;
+			int n = snprintf(&buf[w], sizeof buf - w,
+			                 "\tx%02zx -> x%02x;\n",
+			                 src, dst);
+			if (n < 1)
+				return;
+			w += (size_t)n;
+		}
+	}
+
+	buf[w++] = '}';
+	buf[w++] = '\n';
+	buf[w] = '\0';
+	fputs(buf, stdout);
+}
+#endif
+
+/**
+ * @brief Convert a parser state from a bit flag representation
+ *        to the corresponding parser state enumeration.
+ *
+ * If the return value is not negative, it is guaranteed to be
+ * the `int` conversion of a valid @ref utf8_st8 enumeration.
+ *
+ * @param bit The flag to convert. Must be a non-zero power of 2.
+ * @return An enumeration if `bit` is a valid parser state flag,
+ *         otherwise -1.
+ */
+static const_inline int
+utf8_state_from_bit (uint16_t bit)
+{
+	return bit && !(bit & (bit - 1U))
+		? __builtin_ctz(bit) : -1;
+}
+
+/**
+ * @brief Get the parser's current state as a state enumeration.
+ *
+ * Note that this is not a pure getter, as @ref utf8::error is
+ * clobbered on failure. Take precautions to save it if needed.
+ *
+ * @param u8p The UTF-8 parser. The error field is set on failure
+ *            without regard to its current state.
+ * @param st8 Where to store the state enumeration on success.
+ *            Not modified on failure. Must not be `nullptr`.
+ * @return `true` on success. On failure sets @ref ut8::error to
+ *         `ENOTRECOVERABLE` and returns `false`.
+ */
+nonnull_in()
+static force_inline bool
+utf8_get_state (struct utf8 *const   u8p,
+                enum utf8_st8 *const st8)
+{
+	int e = utf8_state_from_bit(u8p->state);
+	if (e < 0) {
+		u8p->error = ENOTRECOVERABLE;
+		return false;
+	}
+
+	assume_value_bits(e, 0xf);
+
+	*st8 = (enum utf8_st8)e;
+	return true;
+}
+
+/**
+ * @brief Get a mask of state flags comprising the states allowed after `st8`.
+ *
+ * @param st8 A parser state enumeration.
+ * @return Mask of state flags representing the allowed states, or 0 if `st8`
+ *         is not a valid state enumeration.
+ */
+static const_inline uint16_t
+utf8_get_allowed_next_states (enum utf8_st8 st8)
+{
+	if (!is_negative(st8) && st8 <= utf8_ini)
+		return utf8_dst[st8];
+	assert(!is_negative(st8) && st8 <= utf8_ini);
+	return 0;
```

Comment:
This is intentional design. In a release build the `assert()` is a no-op due to `NDEBUG` being defined. In a debug build the `assert()` stays in order to cause program abort if the array index is out of range.

In the debug build the `if` conditional doesn't matter, since a debug build is always expected to have some overhead. The only thing that matters is that the function avoids  out-of-bounds array access with no exceptions; in the debug case the `assert()` will still fire just fine, alerting the developers of a bug somewhere in the calling code.

The worst thing to do would be to use `assert()` as the actual branch point. In my opinion it's not safe to do so even in debugging code, because that makes it more likely that some day an `assert()` will accidentally be left as the only branch mechanism in a production build, too. Not an acceptable risk.

I hope my explanation is understandable, please ask more if there's anything more you need to know to add learnings.

Reply:
Clarifying just one thing: in a debugging build putting the `assert()` below the branch point (instead of above it) means that any conditional instructions executed by the `assert()` will not happen if the array index is valid. So there's that very small optimization too.